File: 671.二叉树中第二小的节点.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int findSecondMinimumValue(TreeNode* root) {
        if(!root)
            return -1;
        if(!root->left && !root->right)
            return -1;
        int left = root->left->val;
        int right = root->right->val;
        if(left == root->val)
            left = findSecondMinimumValue(root->left);
        if(right == root->val)
            right = findSecondMinimumValue(root->right);
        if(left != -1 && right != -1)
            return min(left, right);
        if(left == -1)
            return right;
        else
            return left;
    }
};

File: 671.二叉树中第二小的节点.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var findSecondMinimumValue = function(root) {
    let res = -1;
    const val = root.val;
    const dfs = (root) => {
        if (root === null)
            return;
        if (res !== -1 && root.val >= res)
            return;
        if (root.val > val)
            res = root.val;
        dfs(root.left);
        dfs(root.right);
    }
    dfs(root);
    return res;
};